### PR TITLE
Feature/improved language detection part 2: tracks, saving and clearing language

### PR DESCRIPTION
**Pull request by hafizrahman**
Part of https://github.com/woocommerce/woocommerce-android/issues/9398, which is a subtask of https://github.com/woocommerce/woocommerce-android/issues/9214

Somewhat large PR, but a big chunk of it is a Compose component that's copied from elsewhere, so likely do not need review anymore ([more here](https://github.com/woocommerce/woocommerce-android/pull/9410#pullrequestreview-1528437107)).

### Description

This continues the language detection work that is done to improve AI-generated description and sharing messages. It adds:

1. Saving identified language to be re-used for regeneration.
2. Clearing up saved identified language if user sent negative feedback in the feedback form that appears after generating something.
    - As a subtask of this, the PR also adds feedback form to product sharing bottom sheet. Previously it only appears on the product description generation.
3. Trackings related to language identification.

### Known issue
- The feedback form added in the product sharing bottom sheet made the layout drawn awkwardly when the form is hidden (i.e: after user enters a feedback). I plan to work on the design/animation aspect separately to mimic iOS behavior to make this look less janky. For now I'm prioritizing the functionality first.
- Missing unit tests. Will be added separately too.
- The copied component should probably be refactored away to be reusable. I'll take a look at it separately too.

### Testing instructions

**Testing language saving and the feedback form at the same time:**
This is a bit fuzzy as the result from AI can't be guaranteed, but probably makes things easier to test with a product with name in language A, and product description in language B. The demo product that comes in WooCommerce is a good example to test here as usually the product title is English while the description is in Latin (Lorem ipsum text).

1. Go to the details screen of the above product, tap share on top right,
2. Generate a sharing message, notice the language of the message generated by the AI (can be English or Latin with example above),
3. On the shown feedback form, tap the "thumbs down" button, ensure that it disappears,
4. Regenerate another message. Internally, this will not save the language used in point 2, and the AI sometimes might choose to use the other language. You can retry this over and over until you see this language switch behavior; the switching will indicate that the saving works.

Check tracks on: 

```
🔵 Tracked: ai_identify_language_success, Properties: {"language":"en","source":"product_sharing","blog_id":211272157,"is_wpcom_store":true,"is_debug":true}

🔵 Tracked: product_ai_feedback, Properties: {"source":"product_sharing","is_useful":true,"blog_id":211272157,"is_wpcom_store":true,"is_debug":true}
```

The same behavior will also apply on product description, so feel free to test separately.

### Images/gif
Feedback form shown in product sharing dialog:

[device-2023-07-14-111847.webm](https://github.com/woocommerce/woocommerce-android/assets/266376/6f4d3eae-da4e-4dce-a932-2702a5ce7ae0)

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductSharingBottomSheet.kt [hunk 1]
```diff
@@ -139,6 +151,80 @@ fun ProductShareWithAI(
     }
 }
 
+@Composable
+fun FeedbackForm(onDescriptionFeedbackReceived: (Boolean) -> Unit) {
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductSharingBottomSheet.kt:164]: This composable component is pretty much a copy of an existing compose code on https://github.com/woocommerce/woocommerce-android/blob/99c73aab8f7b9d27e64e9ec6c64fa989d048ff21/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/AIProductDescriptionBottomSheet.kt#L288 

So, it likely dooes not need reviewing anymore.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductSharingViewModel.kt [hunk 2]
```diff
@@ -65,49 +66,102 @@ class ProductSharingViewModel @Inject constructor(
...
     private fun handleCompletionsSuccess(completions: String) {
         tracker.track(AnalyticsEvent.PRODUCT_SHARING_AI_MESSAGE_GENERATED)
 
         _viewState.update {
             it.copy(
                 buttonState = AIButtonState.Regenerate(labelForRegenerate),
                 shareMessage = completions,
-                isGenerating = false
+                isGenerating = false,
+                shouldShowFeedbackForm = !it.isFeedbackReceived
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductSharingViewModel.kt]: This is the main logic about showing/hiding feedback form. 

If completion is successful, then show feedback form if we haven't received feedback from the form (during the lifecycle of this vm, but that's OK).

Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductSharingViewModel.kt]: Resolved as it's been refactored away.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/AIProductDescriptionViewModel.kt [hunk 3]
```diff
@@ -57,30 +61,66 @@ class AIProductDescriptionViewModel @Inject constructor(
...
         launch {
-            generateDescription()
+            _viewState.value.identifiedLanguageISOCode?.let {
+                generateProductDescriptionText(languageISOCode = it)
+            } ?: identifyLanguageAndGenerateProductDescriptionText()
         }
     }
 
-    private suspend fun generateDescription() {
+    private suspend fun identifyLanguageAndGenerateProductDescriptionText() {
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/AIProductDescriptionViewModel.kt]: Np, no need to change. Ideally if we need to name the function using `And` should indicate that we can split the function 2.

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
Great job @hafizrahman 🙌🏼 Everything works as expected and the code looks nice a clean! 

Once the conflicts are resolved: :shipit:

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/AIProductDescriptionViewModel.kt (hunk 3 above)
Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/AIProductDescriptionViewModel.kt]: Thanks! Technically that is already split into two as product description generation is already its own function.

But this comment made me think and I refactored the code a bit to make things clearer with each function specifically doing only one thing.

**Pull request merged**